stan/math/rev/mat/functor/idas_integrator.hpp

```diff
@@ -0,0 +1,330 @@
+ */
+class idas_integrator {
+  const double rtol_;
+  const double atol_;
+  const int64_t max_num_steps_;
+  /**
+   * Forward decl
+   */
+  template <typename Dae>
+  void init_sensitivity(Dae& dae);
+
+  /**
+   * Placeholder for data-only idas_forword_system, no sensitivty
+   *
+   * @tparam F DAE RHS functor type.
+   * @param[out] dae DAE system
+   * @param[in] t0 initial time.
+   * @param[in] ts times of the desired solutions
+   * @param[out] res_yy DAE solutions
+   */
+  template <typename F>
+  void init_sensitivity(idas_forward_system<F, double, double, double>& dae) {}
+
+  /**
+   *  IDAS adjoint sens calculation requires different initialization
+   *
+   * @tparam F type of DAE RHS functor
+   * @tparam Tyy type of DAE primary unknowns
+   * @tparam Typ type of DAE derivative unknowns
+   * @tparam Tpar type of DAE parameters.
+   * @param[out] dae DAE system
+   * @param[in] t0 initial time.
+   * @param[in] ts times of the desired solutions
+   * @param[out] res_yy DAE solutions
+   */
+  template <typename F, typename Tyy, typename Typ, typename Tpar>
+  void init_sensitivity(idas_adjoint_system<F, Tyy, Typ, Tpar>& dae) {
+    // TODO(yizhang): adjoint sensitivity initialization
+  }
+
+  /**
+   * Forward decl
+   */
+  template <typename F>
+  void solve(idas_forward_system<F, double, double, double>& dae,
+             const double& t0, const std::vector<double>& ts,
+             std::vector<std::vector<double> >& res_yy);
+
+  /**
+   * Forward decl
+   */
+  template <typename Dae>
+  void solve(Dae& dae, const double& t0, const std::vector<double>& ts,
+             typename Dae::return_type& res_yy);
+
+  /**
+   * IDAS adjoint system requires different solution process
+   * @tparam F type of DAE RHS functor
+   * @tparam Tyy type of DAE primary unknowns
+   * @tparam Typ type of DAE derivative unknowns
+   * @tparam Tpar type of DAE parameters.
+   * @param[in/out] dae DAE system
+   * @param[in] t0 initial time.
+   * @param[in] ts times of the desired solutions, in strictly
+   * @param[out] res_yy DAE solutions
+   */
+  template <typename F, typename Tyy, typename Typ, typename Tpar>
+  void solve(idas_adjoint_system<F, Tyy, Typ, Tpar>& dae, const double& t0,
+             const std::vector<double>& ts,
+             std::vector<std::vector<double> >& res_yy) {
+    // TODO(yizhang): adjoint sensitivity solver
+  }
+
+ public:
+  static constexpr int IDAS_MAX_STEPS = 500;
+  /**
+   * constructor
+   * @param[in] rtol double relative tolerance.
+   * @param[in] atol double absolute tolerance.
+   * @param[in] max_num_steps max nb. of times steps.
+   */
+  idas_integrator(const double rtol, const double atol,
+                  const int64_t max_num_steps)
+      : rtol_(rtol), atol_(atol), max_num_steps_(max_num_steps) {
+    if (rtol_ <= 0)
+      invalid_argument("idas_integrator", "relative tolerance,", rtol_, "",
+                       ", must be greater than 0");
+    if (rtol_ > 1.0E-3)
```

Comment:
IDAS manual recommends this. Since we don't provide default value here(unlike in ODE solver), we don't want to give the user a foot-shooter, as it may not be easy to spot this in model debugging.